runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcGetDataStream.java

```diff
@@ -350,7 +396,23 @@ private void queueRequestAndWait(QueuedRequest request) throws InterruptedExcept
       batch.countDown();
     } else {
       // Wait for this batch to be sent before parsing the response.
-      batch.await();
+      boolean batchSent = false;
+      long secondsWaited = 0;
+      long waitFor = 10;
+      while (!batchSent) {
+        if (isClosed()) {
```

Comment:
by polling we could add latency in cases where the threads are tied up waiting for reads and thus new work is not processed.

Do we need some finally (batch.countDown()) above to handle exceptions in sendBatch if the stream is closed? In that case it seems like it woudl unspool without needing this isClosed check here.

Reply:
messaged offline